File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l,r=0,len(nums)-1
        res = nums[0]
        while l<=r:
            if nums[l]<nums[r]:
                res = min(res, nums[l])
                break

            mid = (l+r)//2
            res = min(res, nums[mid])

            if nums[mid]>=nums[l]:
                l = mid+1
            else:
                r = mid-1

        return res
            # if mid is greater than first element and first element is greater than last element
            #min must be in the right half
            # if [6,1,2,3,4,5]
            #if mid is smaller than first element search in the left half

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> preproduct (nums.size());
        vector<int> postproduct (nums.size());
        int preP = 1, postP = 1;

        for (int i=0;i<nums.size()-1;i++){
            preP *= nums[i];
            preproduct[i] = preP;
        }

        for(int i=nums.size()-1;i>=0;i--){
            postP*=nums[i];
            postproduct[i] = postP;
        }
        
        vector<int> res;
        for(int i=0;i<=nums.size()-1;i++){
            if(i==0){
                //first element
                res.push_back(postproduct[i+1]);
            }
            else if(i==nums.size()-1){
                //last element
                res.push_back(preproduct[i-1]);
            }
            else{
                int prod = preproduct[i-1]*postproduct[i+1];
                res.push_back(prod);
            }
        }
        return res;
    }
};
